rust-runtime/smithy-eventstream/src/frame.rs

```diff
@@ -0,0 +1,590 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Event Stream message frame types and serialization/deserialization logic.
+
+use crate::buf::count::CountBuf;
+use crate::buf::crc::{CrcBuf, CrcBufMut};
+use crate::error::Error;
+use crate::str_bytes::StrBytes;
+use bytes::{Buf, BufMut, Bytes};
+use std::convert::TryFrom;
+use std::mem::size_of;
+
+const PRELUDE_LENGTH_BYTES: u32 = 3 * size_of::<u32>() as u32;
+const PRELUDE_LENGTH_BYTES_USIZE: usize = PRELUDE_LENGTH_BYTES as usize;
+const MESSAGE_CRC_LENGTH_BYTES: u32 = size_of::<u32>() as u32;
+const MAX_HEADER_NAME_LEN: usize = 255;
+const MIN_HEADER_LEN: usize = 2;
+
+mod value {
+    use crate::error::Error;
+    use crate::frame::checked;
+    use crate::str_bytes::StrBytes;
+    use bytes::{Buf, BufMut, Bytes};
+    use smithy_types::Instant;
+    use std::mem::size_of;
+
+    const TYPE_TRUE: u8 = 0;
+    const TYPE_FALSE: u8 = 1;
+    const TYPE_BYTE: u8 = 2;
+    const TYPE_INT16: u8 = 3;
+    const TYPE_INT32: u8 = 4;
+    const TYPE_INT64: u8 = 5;
+    const TYPE_BYTE_ARRAY: u8 = 6;
+    const TYPE_STRING: u8 = 7;
+    const TYPE_TIMESTAMP: u8 = 8;
+    const TYPE_UUID: u8 = 9;
+
+    /// Event Stream frame header value.
+    #[non_exhaustive]
+    #[derive(Clone, Debug, PartialEq)]
+    pub enum HeaderValue {
+        Bool(bool),
+        Byte(u8),
+        Int16(i16),
+        Int32(i32),
+        Int64(i64),
+        ByteArray(Bytes),
+        String(StrBytes),
```

Comment:
if `StrBytes` pre-validates (as it probably should, then users don't need another layer of checking here)